File: lua/utils/theme_sample.ts
// Theme Preview Sample

const TIMEOUT = 3000
const API_URL = "https://api.example.com"

interface Animal {
  kind: string
  sound: string
}

type FetchResult<T> = {
  status: number
  data: T | null
  error?: string
}

class Dog implements Animal {
  kind = "dog"
  sound = "woof"

  speak(): string {
    return `The ${this.kind} says: ${this.sound}`
  }
}

function greet(name: string, opts?: { loud?: boolean }): string {
  if (!name) throw new Error("name cannot be empty")

  const msg = `Hello, ${name}!`
  return opts?.loud ? msg.toUpperCase() : msg
}

async function fetchData<T>(url: string): Promise<FetchResult<T>> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), TIMEOUT)

  try {
    const res = await fetch(url, { signal: controller.signal })

    if (!res.ok) {
      return { status: res.status, data: null, error: res.statusText }
    }

    const data = (await res.json()) as T
    return { status: res.status, data }
  } catch (err) {
    return { status: 0, data: null, error: String(err) }
  } finally {
    clearTimeout(timer)
  }
}

const items = ["apple", "banana", "cherry"]

const formatted = items
  .map((item) => item[0].toUpperCase() + item.slice(1))
  .filter((item) => item.length > 5)

// TODO: add pagination support
// FIXME: error boundary missing
async function main() {
  const dog = new Dog()
  console.log(dog.speak())
  console.log(greet("world", { loud: true }))

  const result = await fetchData<{ items: string[] }>(API_URL + "/data")

  if (result.error) {
    console.error("Error:", result.error)
    return
  }

  console.log("Items:", formatted)
}

main()

File: lua/utils/theme_sample.go
package main

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"time"
)

const Timeout = 3 * time.Second
const BaseURL = "https://api.example.com"

type Animal struct {
	Kind  string
	Sound string
}

func (a Animal) Speak() string {
	return fmt.Sprintf("The %s says: %s", a.Kind, a.Sound)
}

type fetchResult struct {
	Status int
	Body   string
}

func fetch(ctx context.Context, url string) (*fetchResult, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("creating request: %w", err)
	}

	client := &http.Client{Timeout: Timeout}
	resp, err := client.Do(req)
	if err != nil {
		return nil, errors.New("request failed")
	}
	defer resp.Body.Close()

	return &fetchResult{Status: resp.StatusCode, Body: "ok"}, nil
}

func greet(name string, loud bool) string {
	if name == "" {
		panic("name cannot be empty")
	}

	msg := fmt.Sprintf("Hello, %s!", name)

	if loud {
		return fmt.Sprintf("%q", msg)
	}

	return msg
}

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), Timeout)
	defer cancel()

	items := []string{"apple", "banana", "cherry"}

	for i, item := range items {
		items[i] = fmt.Sprintf("%s!", item)
	}

	dog := Animal{Kind: "dog", Sound: "woof"}
	fmt.Println(dog.Speak())
	fmt.Println(greet("world", true))

	// TODO: handle pagination
	// FIXME: retry logic missing
	result, err := fetch(ctx, BaseURL+"/data")
	if err != nil {
		fmt.Println("error:", err)
		return
	}

	fmt.Println("status:", result.Status)
}
